File: Server~/src/tools/textureTools.ts
import * as z from 'zod';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { getToolTimeout } from '../utils/timeouts.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// ============================================================================
// configure_texture_settings
// ============================================================================

const toolName = 'configure_texture_settings';
const toolDescription = 'Configures texture asset import settings (Normal Map, Sprite, Default, Max Size, sRGB, Filter Mode, Wrap Mode, Read/Write).';
const paramsSchema = z.object({
  texturePath: z.string().describe('Project path or GUID of the texture asset to configure (e.g. "Assets/Textures/Stone_Normal.png")'),
  textureType: z.enum(['default', 'normalmap', 'sprite', 'cookie', 'gui', 'singlechannel']).optional().describe('Texture type interpretation'),
  maxTextureSize: z.number().int().optional().describe('Maximum allowed texture resolution (e.g. 256, 512, 1024, 2048, 4096)'),
  wrapMode: z.enum(['repeat', 'clamp', 'mirror', 'mirroronce']).optional().describe('Texture wrap/tiling mode'),
  filterMode: z.enum(['point', 'bilinear', 'trilinear']).optional().describe('Filtering mode (Point for pixel art, Bilinear/Trilinear for smooth rendering)'),
  sRGBTexture: z.boolean().optional().describe('Enable for color textures (Albedo/Diffuse), disable for linear data textures (Normal Maps, Roughness, Masks)'),
  isReadable: z.boolean().optional().describe('Enable Read/Write on CPU for texture pixel sampling'),
  generateMipMaps: z.boolean().optional().describe('Whether to generate mipmap levels for distant rendering'),
  compression: z.enum(['compressed', 'uncompressed', 'highquality', 'lowquality']).optional().describe('Texture compression format quality'),
  reason: z.string().optional().describe('Optional explanation of why texture settings are being modified')
});

export function registerConfigureTextureSettingsTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: z.infer<typeof paramsSchema>) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: z.infer<typeof paramsSchema>): Promise<CallToolResult> {
  if (!params.texturePath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Parameter 'texturePath' must be provided."
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  }, { timeout: getToolTimeout(toolName) });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `Failed to configure texture settings`
    );
  }

  return {
    content: [{
      type: response.type || 'text',
      text: response.message || `Successfully configured texture settings`
    }],
    structuredContent: {
      texturePath: response.texturePath,
      textureType: response.textureType,
      maxTextureSize: response.maxTextureSize,
      wrapMode: response.wrapMode,
      filterMode: response.filterMode,
      sRGBTexture: response.sRGBTexture,
      isReadable: response.isReadable,
      mipmapEnabled: response.mipmapEnabled
    }
  };
}

File: Editor/Tools/TextureTools.cs
using System;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for configuring texture asset import settings (TextureType, NormalMap, Sprite, max size, sRGB, Read/Write).
    /// </summary>
    public class ConfigureTextureSettingsTool : McpToolBase
    {
        public ConfigureTextureSettingsTool()
        {
            Name = "configure_texture_settings";
            Description = "Configures texture asset import settings (Normal Map, Sprite, Default, Max Size, sRGB, Filter Mode, Wrap Mode, Read/Write).";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            string texturePath = parameters["texturePath"]?.ToObject<string>();
            string textureTypeStr = parameters["textureType"]?.ToObject<string>()?.ToLowerInvariant();
            int? maxSize = parameters["maxTextureSize"]?.ToObject<int?>();
            string wrapModeStr = parameters["wrapMode"]?.ToObject<string>()?.ToLowerInvariant();
            string filterModeStr = parameters["filterMode"]?.ToObject<string>()?.ToLowerInvariant();
            bool? sRGB = parameters["sRGBTexture"]?.ToObject<bool?>();
            bool? isReadable = parameters["isReadable"]?.ToObject<bool?>();
            bool? generateMipMaps = parameters["generateMipMaps"]?.ToObject<bool?>();
            string compressionStr = parameters["compression"]?.ToObject<string>()?.ToLowerInvariant();
            string reason = parameters["reason"]?.ToObject<string>();

            if (string.IsNullOrWhiteSpace(texturePath))
            {
                return McpUnitySocketHandler.CreateErrorResponse("Parameter 'texturePath' is required.", "validation_error");
            }

            // Resolve path
            texturePath = texturePath.Replace('\\', '/').Trim();
            if (!texturePath.StartsWith("Assets/") && !texturePath.StartsWith("Packages/"))
            {
                string[] guids = AssetDatabase.FindAssets($"{texturePath} t:Texture2D");
                if (guids.Length > 0)
                {
                    texturePath = AssetDatabase.GUIDToAssetPath(guids[0]);
                }
                else
                {
                    texturePath = "Assets/" + texturePath.TrimStart('/');
                }
            }

            TextureImporter importer = AssetImporter.GetAtPath(texturePath) as TextureImporter;
            if (importer == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse($"Texture asset or TextureImporter not found at path '{texturePath}'.", "asset_not_found");
            }

            // Texture type
            if (!string.IsNullOrEmpty(textureTypeStr))
            {
                switch (textureTypeStr)
                {
                    case "normalmap":
                    case "normal_map":
                    case "normal":
                        importer.textureType = TextureImporterType.NormalMap;
                        importer.sRGBTexture = false;
                        break;
                    case "sprite":
                    case "sprite2d":
                        importer.textureType = TextureImporterType.Sprite;
                        break;
                    case "cookie":
                        importer.textureType = TextureImporterType.Cookie;
                        break;
                    case "gui":
                    case "ui":
                        importer.textureType = TextureImporterType.GUI;
                        break;
                    case "singlechannel":
                    case "mask":
                        importer.textureType = TextureImporterType.SingleChannel;
                        importer.sRGBTexture = false;
                        break;
                    case "default":
                    default:
                        importer.textureType = TextureImporterType.Default;
                        break;
                }
            }

            if (maxSize.HasValue)
            {
                importer.maxTextureSize = Mathf.ClosestPowerOfTwo(maxSize.Value);
            }

            if (sRGB.HasValue)
            {
                importer.sRGBTexture = sRGB.Value;
            }

            if (isReadable.HasValue)
            {
                importer.isReadable = isReadable.Value;
            }

            if (generateMipMaps.HasValue)
            {
                importer.mipmapEnabled = generateMipMaps.Value;
            }

            if (!string.IsNullOrEmpty(wrapModeStr))
            {
                switch (wrapModeStr)
                {
                    case "clamp": importer.wrapMode = TextureWrapMode.Clamp; break;
                    case "mirror": importer.wrapMode = TextureWrapMode.Mirror; break;
                    case "mirroronce": importer.wrapMode = TextureWrapMode.MirrorOnce; break;
                    case "repeat":
                    default: importer.wrapMode = TextureWrapMode.Repeat; break;
                }
            }

            if (!string.IsNullOrEmpty(filterModeStr))
            {
                switch (filterModeStr)
                {
                    case "point": importer.filterMode = FilterMode.Point; break;
                    case "trilinear": importer.filterMode = FilterMode.Trilinear; break;
                    case "bilinear":
                    default: importer.filterMode = FilterMode.Bilinear; break;
                }
            }

            if (!string.IsNullOrEmpty(compressionStr))
            {
                switch (compressionStr)
                {
                    case "uncompressed":
                    case "none":
                        importer.textureCompression = TextureImporterCompression.Uncompressed;
                        break;
                    case "highquality":
                    case "hq":
                        importer.textureCompression = TextureImporterCompression.CompressedHQ;
                        break;
                    case "lowquality":
                    case "lq":
                        importer.textureCompression = TextureImporterCompression.CompressedLQ;
                        break;
                    case "compressed":
                    default:
                        importer.textureCompression = TextureImporterCompression.Compressed;
                        break;
                }
            }

            importer.SaveAndReimport();

            McpLogger.LogInfo($"[MCP Unity] Configured TextureImporter for '{texturePath}' (type: {importer.textureType}, maxSize: {importer.maxTextureSize})" + (reason != null ? $" — {reason}" : ""));

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Successfully updated texture import settings for '{texturePath}'",
                ["texturePath"] = texturePath,
                ["textureType"] = importer.textureType.ToString(),
                ["maxTextureSize"] = importer.maxTextureSize,
                ["wrapMode"] = importer.wrapMode.ToString(),
                ["filterMode"] = importer.filterMode.ToString(),
                ["sRGBTexture"] = importer.sRGBTexture,
                ["isReadable"] = importer.isReadable,
                ["mipmapEnabled"] = importer.mipmapEnabled
            };
        }
    }
}
